pkg/collector/smi_collector.go

```diff
@@ -0,0 +1,99 @@
+package collector
+
+import (
+	"fmt"
+	"log"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/aks-periscope/pkg/interfaces"
+	"github.com/Azure/aks-periscope/pkg/utils"
+)
+
+// SmiCollector defines an Smi Collector struct
+type SmiCollector struct {
+	BaseCollector
+}
+
+var _ interfaces.Collector = &SmiCollector{}
+
+// NewSmiCollector is a constructor
+func NewSmiCollector(exporter interfaces.Exporter) *SmiCollector {
+	return &SmiCollector{
+		BaseCollector: BaseCollector{
+			collectorType: Smi,
+			exporter:      exporter,
+		},
+	}
+}
+
+// Collect implements the interface method
+func (collector *SmiCollector) Collect() error {
+	// Get all CustomResourceDefinitions in the cluster
+	allCrdsList, err := utils.GetResourceList([]string{"get", "crds", "-o", "jsonpath={..metadata.name}"}, " ")
+	if err != nil {
+		return err
+	}
+
+	// Directory where logs will be written to
+	rootPath, err := utils.CreateCollectorDir(collector.GetName())
+	if err != nil {
+		return err
+	}
+
+	// Filter to obtain a list of Smi CustomResourceDefinitions in the cluster
+	crdNameContainsSmiTestPredicate := func(s string) bool { return strings.Contains(s, "smi-spec.io") }
+	smiCrdsList := utils.FilterSliceElemsWithTestPredicate(allCrdsList, crdNameContainsSmiTestPredicate)
+	if len(smiCrdsList) == 0 {
+		return fmt.Errorf("Cluster does not contain any SMI CRDs")
+	}
+
+	collectSmiCrds(collector, filepath.Join(rootPath, "smi_crd_definitions"), smiCrdsList)
+	return collectSmiCustomResourcesFromAllNamespaces(collector, filepath.Join(rootPath, "smi_custom_resources"), smiCrdsList)
+}
+
+func collectSmiCrds(collector *SmiCollector, rootPath string, smiCrdsList []string) {
+	for _, smiCrd := range smiCrdsList {
+		fileName := smiCrd + "_definition.yaml"
+		kubeCmd := []string{"get", "crd", smiCrd, "-o", "yaml"}
+		if err := collector.CollectKubectlOutputToCollectorFiles(rootPath, fileName, kubeCmd); err != nil {
+			log.Printf("Skipping: unable to collect yaml definition of %s to %s: %+v", smiCrd, fileName, err)
+		}
+	}
+}
+
+func collectSmiCustomResourcesFromAllNamespaces(collector *SmiCollector, rootPath string, smiCrdsList []string) error {
+	// Get all namespaces in the cluster
+	namespacesList, err := utils.GetResourceList([]string{"get", "namespaces", "-o", "jsonpath={..metadata.name}"}, " ")
+	if err != nil {
+		return fmt.Errorf("Failed to collect SMI custom resources: unable to list namespaces in the cluster: %+v", err)
```

Comment:
This line produces a warning:
- no capital letter
- %w instead of %v
- as you are already in an error message you could simplify the text

`return fmt.Errorf("collect SMI custom resources: unable to list namespaces in the cluster: %w", err)`